// shipengine/dist/models/package.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// TODO change name to parcel to not conflict with JS keyword
var Package =
// TODO add label messages
// https://docs.shipengine.com/docs/custom-label-messages
/**
   *
   * @param {Object} weight - Object with description of package weight
   * @param {number} weight.value - decimal number value of package weight
   * @param {string} weight.unit - unit of measurment
   * @param {string} package_code
   * @param {Object} [dimensions=null] - Physical dimensions of the package to ship
   * @param {Object} [dimensions.unit] - Unit of measurement for package dimensions
   * @param {number} [dimensions.length] - Decimal number for length of package to ship
   * @param {number} [dimensions.width] - Decimal number for width of package to ship
   * @param {number} [dimensions.height] - Deciaml number for height of package to ship
   * @param {number} [insured_value=null] - Decimal number (in USD) of amount to insure package for
   */
function Package(weight) {
  var dimensions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  var package_code = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var insured_value = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;

  _classCallCheck(this, Package);

  this.weight = weight;

  if (package_code) this.package_code = package_code;
  if (dimensions) this.dimensions = dimensions;

  if (insured_value !== null) {
    this.insured_value = {
      currency: 'usd',
      amount: insured_value
    };
  }
};

exports.Package = Package;
